fix(core/dfn-index): escape HTML/XML tags

nigelmegitt opened the pull request:
A fix for #4430 : when the normalized text is rendered as html e.g. in `dfn-index.js` line 298 within `renderExternalTermEntry` using `html: text` this prevents creation of what the UA thinks are HTML elements from terms defined with angle brackets at the start and end.

An alternative fix could be to replace `html: text` with `text: text` but I think that would break cases where code blocks are deliberately put around the term in the index.

This fix seems to pass all the tests.

---

- Fixes #4430
- Fixes https://github.com/w3c/respec/issues/4324

nigelmegitt commented:
Ideally I would have added a test for this but I wasn't sure the best way to do that. Hints accepted!

sidvishnoi commented:
In terms of new test, it'd be be great if you can add the `<ttp:feature>` reference mentioned in #4430: https://github.com/w3c/respec/blob/a950478bca8bcfc3f540ca8d40a4e5e01ae5cd7d/tests/spec/core/dfn-index-spec.js#L159

---

Updating the `norm` function here seems a bit risky to me, even though the tests pass. `norm` is used in a lot more places, that might not be tested well enough.

Can we just update the part where this is rendered (consider adding the test first):
https://github.com/w3c/respec/blob/a950478bca8bcfc3f540ca8d40a4e5e01ae5cd7d/src/core/dfn-index.js#L338

sidvishnoi commented:
Actually, if you use following, we get fix for #4324 too:
```diff
diff --git a/src/core/xref.js b/src/core/xref.js
index 4e0e9e8b..e53048e0 100644
--- a/src/core/xref.js
+++ b/src/core/xref.js
@@ -25,6 +25,7 @@ import {
   nonNormativeSelector,
   norm as normalize,
   showError,
+  xmlEscape,
 } from "./utils.js";
 import { possibleExternalLinks } from "./link-to-dfn.js";
 import { sub } from "./pubsubhub.js";
@@ -189,7 +190,7 @@ function getRequestEntry(elem) {
 export function getTermFromElement(elem) {
   const { lt: linkingText } = elem.dataset;
   let term = linkingText ? linkingText.split("|", 1)[0] : elem.textContent;
-  term = normalize(term);
+  term = xmlEscape(normalize(term));
   return term === "the-empty-string" ? "" : term;
 }
```

as `getTermFromElement` is used in both places (and only in these two places).

nigelmegitt commented:
Thanks @sidvishnoi, trying that now...

nigelmegitt commented:
@sidvishnoi thanks for the tips, that appears to work. I reverted my change, created the tests first, verified that they failed, implemented your proposal, and verified that the tests passed and that with an example document the behaviour was indeed as expected. 

I'm not sure how to write the test for #4324 because it seems to rely on a behaviour within the Respec "pill" specifically, so I cannot confirm that this does fix it, however it seems likely that it would.

sidvishnoi commented:
Let's skip tests for #4324, I verified locally.

sidvishnoi commented:
Released in [`v34.0.1`](https://github.com/w3c/respec/releases/tag/v34.0.1).
